public/app/plugins/datasource/loki/components/LokiContextUi.tsx

```diff
@@ -143,8 +171,44 @@ export function LokiContextUi(props: LokiContextUiProps) {
 
   useAsync(async () => {
     setLoading(true);
-    const contextFilters = await logContextProvider.getInitContextFiltersFromLabels(row.labels, origQuery);
-    setContextFilters(contextFilters);
+    const initContextFilters = await logContextProvider.getInitContextFiltersFromLabels(row.labels, origQuery);
+
+    let preservedLabels: undefined | PreservedLabels = undefined;
+    try {
+      preservedLabels = JSON.parse(store.get(LOKI_LOG_CONTEXT_PRESERVED_LABELS));
+      // Do nothing when error occurs
+    } catch (e) {}
+
+    if (!preservedLabels) {
+      setContextFilters(initContextFilters);
+    } else {
+      // We need to update filters based on preserved labels
+      let arePreservedLabelsUsed = false;
+      const newContextFilters = initContextFilters.map((contextFilter) => {
+        if (preservedLabels && preservedLabels.removedLabels.includes(contextFilter.label)) {
```

Comment:
```suggestion
        if (preservedLabels.removedLabels.includes(contextFilter.label)) {
```

Should be covered by line 182 anyways.

Reply:
typescript is a little bit dumb here 

<img width="687" alt="image" src="https://github.com/grafana/grafana/assets/30407135/0d543535-a208-48af-b7e2-801fbd10abc5">

But I changed it to use `!` as we know at that point that we have preserved labels https://github.com/grafana/grafana/pull/68700/commits/6c8a00bed3ddcee16b901b07842eb1653d6c5b41